### DISCUSSION TITLE
xDrip Follower Sync Network Details

### DISCUSSION THREAD

**Discussion by rpm5099** [NightscoutFoundation/xDrip#3370] [Bug reports]
At this point I'm not reporting a specific bug, I just am looking for the details on what communication method the xdrip follower sync uses across the network.  I'm not seeing anything about this in the documentation, and searching the source code did not quickly provide me with an answer.  I'm spending a lot of time trying to make uninformed guesses as to why the xdrip follower sync keeps dropping out.  In making some changes to my network (which is quite complex, but entirely controlled by me) I noticed that the sync stopped working but I've not been able to nail down exactly why.  The master device is set to upload to nightscout as well as be a master for xdrip follower sync devices.  Nightscout uploads are uninterrupted (as I have complete control over that).  I think I saw somewhere that follower sync uses google API's for messaging and I see some log data referencing firebase - does xdrip follower sync send data via google API's?  Does it use the google account active on the phone to access the API's, or does it use a dedicated xrip one, or is that service free?  

More directly, **what ports, protocols, domains, ip addresses, accounts, api keys, country geolocation, etc are required for xdrip follower sync?  Is this configuration the same for both master and follower?  What are the possible/usual/likely causes of xdrip follower sync communication failure?**

Slightly related note - how do I export the raw logs that I see in "view events log".  There seems to be an option to email jamorham the logs and enter your own email, but it does nothing.  Also, the logs are not saved anywhere you can access on the device normal USB.  This shouldn't be something you need root to access.

Thanks in advance.

**Comment by Navid200** [comment]
It uses Google messaging.
You don't need a Google account.  But, you need Google Play service to be installed on the phone.  On most phones it is by default.

We have it on our to-do list to add the ability to save the logs locally.

**Comment by rpm5099** [comment]
I assume you are talking about the [MessageApi](https://developers.google.com/android/reference/com/google/android/gms/wearable/MessageApi)?  I was able to figure it out the issue after looking for traffic going to/from the google play api, it was not related to xdrip, although it would have been helpful to be able to see more detailed logs regarding the success/failure status of the MessageApi commands.  Is there a list anywhere of the "extra logging tags"?  

That's good news about the local log saving - keep up the good work, it is greatly appreciated!

**Reply by Navid200** [reply]
xDrip offers detail logs if you enable them.
This is how it is done for one particular function (method): https://github.com/NightscoutFoundation/xDrip/wiki/Enabling-Debugging-for-OB1-collector

**Reply by rpm5099** [reply]
Yes, I saw that and I do have the detail logs enabled for the OBD collector, but that is unrelated to network communications.  I was asking where there is a full list of valid extra logging tags - I assume that these two are not the only ones. 
 `Ob1G5CollectionService:v,Ob1G5StateMachine:v`.
 
 Also, to answer my own question in case anyone else is troubleshooting, follower sync network info (at least in the US) goes to `instantmessaging-pa-jms-us.googleapis.com` on port 443 TCP, and looks like normal SSL.

*Edit: * FYI - looks like there might be several subdomains to `googleapis.com` involved.  Difficult to tell which ones are specifically from xdrip from network logs.

**Reply by Navid200** [reply]
You can make any of the Java methods verbose with respect to all their logs.

If there is a Java method titled abc.java, you will add abc:v to get its detail logs.

**Reply by rpm5099** [reply]
Ok great thanks, I'll give it a shot.
